docs/proposal-new-output-syntax.md

```diff
@@ -0,0 +1,705 @@
+  branching point)
+```
+# OK
+a?
+a?fail
+```
+
+- however it is OK to have one of `succeed` and `fail` expected and one
+  optional. This allows triggering off of the not-expected output at the same
+  time as the task is flagged as incomplete
+```
+# OK
+a
+a?fail => call_the_police
+```
+(Note this means we can't have *"`a` means `a:succeed` unless declared optional
+elsewhere"*.)
+
+### Custom Outputs
+
+Custom outputs must be either expected or optional, they can't be both
+```
+# ERROR
+a:x
+a?x
+```
+
+### QUESTION: submit-fail, submit, and start?
+
+It makes some sense to allow the `fail` output to be expected, but allowing
+`submit-fail` to be expected means treating a task as incomplete if it
+successfully submits. The only use case I've thought of
+is to make it somewhat easier (less cleanup before shutdown) to write
+functional tests that deliberately submit to a fake platform.
+
+Similarly, allowing `submit` and `start` to be optional equates to calling a
+task complete even though it didn't run at all, which seems equally perverse.
+
+```
+a:submit-fail  # ERROR or OK?
+a?submit  # ERROR or OK?
+a?start  # ERROR or OK?
+```
+
+## Path Branching
+
+### Concurrent Paths 
+
+Concurrent paths can trigger off of expected outputs:
+```
+a:x => b1
+a:y => b2
+b1 & b2 => c  # join
+```
+
+### Alternate Paths
+
+Alternate paths must branch off of optional outputs, and must be joined with
+OR triggers because only one side or the other will run:
+
+```
+a?x => b1
+a?y => b2
+b1 | b2 => c  # join
+```
+Note joining can trigger off of expected outputs, because the output is
+only "expected" if the task actually runs.
+
+**Failure recovery** is a common example of this:
+```
+a? => b1  # success path
+a?fail => b2  # failure path
+b1 | b2 => c  # join
+```
+
+Note that an artificial dependency can be used to **ensure that at least one
+of the alternate paths gets taken**:
+```
+a?x => b1
+a?y => b2
+b1 | b2 => c  # join
+a => c # ARTIFICIAL DEPENDENCY
+```
+The artificial trigger forces the unsatisfied prerequisites of `c` to be
+spawned, as waiting on `b1` or `b2` even if neither of the optional outputs 
+get completed. If the alternate paths don't actually need to join, `c` could be
+a dummy task.
+
+### QUESTION: Output Completion Expressions?
+
+In the previous example, an artificial dependency (possibly triggering a
+dummy task) is used to deliberately spawn an unsatisfied prerequisite in order
+to ensure that at least one of the alternate paths gets taken. However, that is
+not an ideal way to alert users to a problem:
+- the need to use an artificial dependency won't be obvious to users
+- an unsatisfied prerequisite can't be definitively identified as an error
+  until/unless the workflow stalls
+- and it is an indirect indicator: it does not highlight the actual task
+  that caused the problem
+
+An incomplete task is a direct and immediate indicator. So maybe we can allow
+users to say exactly when a task can be considered complete:
```

Comment:
This looks complicated and only helps if you're dealing with multiple outputs from the same task.
For example:
```
a => b1? & b2?
b1? | b2? => c  # join
a => c # ARTIFICIAL DEPENDENCY
```
In this case the artificial dependency is still needed to ensure that either `b1` or `b2` succeed.

Reply:
OK, I'll move that section to an appendix, as a possible follow-up.  And I'll note your example there too.

Reply:
Moved to an appendix as a possible follow-up.  But on reflection I don't think your example makes much sense - see the new "caveat?" sub-section in the proposal.